### PR TITLE
chore: ensures axe Pro api

### PR DESCRIPTION
**Pull request by AutoSponge**
This adds tests that will help prevent breaking changes in axe-core that axe Pro depends on.

Closes issue: N/A (Q3 rock)

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**

- [x] Follows the commit message policy, appropriate for next version
- [x] Code is reviewed for security

### CODE REVIEW

File: test/integration/api/external/index.js [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+	document.createElement('article')
+];
+
+var inTree = [];
+var walker = collectNodes();
+var next = walker.iterate().next();
+while (!next.done) {
+	inTree.push(next.value);
+	next = walker.iterate().next();
+}
```
Comment (AutoSponge) [inline test/integration/api/external/index.js]: This worked as a spreadable array until I realized we can't use Symbol or rest/spread.

**Comment by straker** [comment]
Question: Most of these checks are just checking that the API exists, is a function, and returns the correct thing. Should these be unit tests in the respected files rather than an integration test all in one?

File: test/integration/api/external/index.js [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+describe('external API', function() {
+	'use strict';
+
+	afterEach(function() {
+		// setup _tree as needed, always reset
+		axe._tree = null;
+	});
+
+	describe('axe.commons.text.sanitize', function() {
+		it('typeof fn', function() {
```
Comment (stephenmathieson) [inline test/integration/api/external/index.js]: What about `it('is a function')`? The test name should be human-readable IMO.

File: test/integration/api/external/index.js [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+			assert.isFunction(axe.commons.text.sanitize);
```
Comment (stephenmathieson) [inline test/integration/api/external/index.js]: Is this test actually useful?

File: test/integration/api/external/index.js [hunk 4]
```diff
@@ -0,0 +1,245 @@
...
+		});
+		it('str => str', function() {
```
Comment (stephenmathieson) [inline test/integration/api/external/index.js]: What?

File: test/integration/api/external/index.js [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+			assert.isArray(axe.utils.validLangs(document));
+			assert.isArray(axe.utils.validLangs({}));
+		});
+	});
+
+	describe('axe.commons.dom.isVisible', function() {
+		it('typeof fn', function() {
+			assert.isFunction(axe.commons.dom.isVisible);
+		});
+		it('el => bool', function() {
```
Comment (stephenmathieson) [inline test/integration/api/external/index.js]: Maybe: `it('returns a boolean when given an element reference')`?

File: test/integration/api/external/index.js [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+			var el = randomNodeInTree(isElement);
+			console.log(el);
```
Comment (stephenmathieson) [inline test/integration/api/external/index.js]: `console.log()`

File: test/integration/api/external/index.js [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+			var role = axe.commons.aria.implicitRole(randomNodeInTree());
+			assert.oneOf(role, implicitRolesOrNull);
+		});
+	});
+
+	describe('axe.commons.aria.lookupTable.role', function() {
+		it('typeof obj', function() {
+			assert.isObject(axe.commons.aria.lookupTable.role);
+		});
+		it('str <=> str (role.type)', function() {
```
Comment (stephenmathieson) [inline test/integration/api/external/index.js]: I have no idea what this means.

File: test/integration/api/external/index.js [hunk 8]
```diff
@@ -0,0 +1,245 @@
...
+	});
+
+	describe('axe.utils.getFlattenedTree', function() {
+		it('typeof fn', function() {
+			assert.isFunction(axe.utils.getFlattenedTree);
+		});
+		it('el => arr', function() {
+			assert.isArray(axe.utils.getFlattenedTree(document.body));
+		});
+		it('el => [{root}]', function() {
```
Comment (stephenmathieson) [inline test/integration/api/external/index.js]: What is this?

File: test/integration/api/external/index.js [hunk 9]
```diff
@@ -0,0 +1,245 @@
...
+			assert.isFunction(axe.utils.getNodeFromTree);
+		});
+		it('node => vnode', function() {
+			axe._tree = axe.utils.getFlattenedTree(document.body);
+			assert.oneOf(
+				axe.utils.getNodeFromTree(randomNodeInTree()),
+				flat(axe._tree[0])
+			);
+		});
+		it('!node => ()', function() {
```
Comment (stephenmathieson) [inline test/integration/api/external/index.js]: I have no idea what this means.

File: test/integration/api/external/index.js [hunk 10]
```diff
@@ -0,0 +1,245 @@
...
+			assert.isUndefined(axe.utils.getNodeFromTree(randomElement()));
+		});
+	});
+
+	describe('axe.commons.dom.isOpaque', function() {
+		it('typeof fn', function() {
+			assert.isFunction(axe.commons.dom.isOpaque);
+		});
+		it('el inTree => bool', function() {
```
Comment (stephenmathieson) [inline test/integration/api/external/index.js]: More confusing test names.

**Comment by AutoSponge** [comment]
> Question: Most of these checks are just checking that the API exists, is a function, and returns the correct thing. Should these be unit tests in the respected files rather than an integration test all in one?

This file should be marked for special review once it lands, the same way we protect manifest. Changes in this file *only* could represent a shift in the axe-core api that breaks axe Pro. For that reason, spreading these tests into their respective modules would make it harder to keep an "eye" on the api axe Pro requires to operate.

File: test/integration/api/external/index.js (hunk 3 above)
Comment (stephenmathieson) [reply test/integration/api/external/index.js]: I don't think we need this test, as the test below it requires `axe.commons.sanitize` to be a function in order to pass.

**Comment by WilcoFiers** [comment]
@AutoSponge +1 on the decision to only test call signatures. We're testing those functions in depth elsewhere. I agree with the idea of adding you and/or Harris as code owner for these tests.

File: build/test/config.js [hunk 11]
```diff
@@ -26,7 +26,10 @@ exports = module.exports = function(grunt, options) {
 					'http://' + host + ':<%= connect.test.options.port %>/test/commons/',
 					'http://' +
 						host +
-						':<%= connect.test.options.port %>/test/integration/rules/'
+						':<%= connect.test.options.port %>/test/integration/rules/',
+					'http://' +
+						host +
+						':<%= connect.test.options.port %>/test/integration/api/external/'
```
Comment (jkodu) [inline build/test/config.js]: I like the new directory `/api`, what is the thinking behind `/external`?

File: test/integration/api/external/index.html [hunk 12]
```diff
@@ -0,0 +1,28 @@
+<!DOCTYPE html PUBLIC "-//W3C//DTD HTML+ARIA 1.0//EN" "http://www.w3.org/WAI/ARIA/schemata/html4-aria-1.dtd">
```
Comment (jkodu) [inline test/integration/api/external/index.html]: enlighten me as to why such verbosity in doctype is essential for these tests?

Comment (AutoSponge) [reply test/integration/api/external/index.html]: This was copied from other axe-core test harnesses. It's not essential to this test.

File: build/test/config.js (hunk 11 above)
Comment (AutoSponge) [reply build/test/config.js]: In the future, someone may want to document the API of a submodule. It also makes it clear this only refers to the API surface area with external dependents.

**Comment by stephenmathieson** [comment]
Is there anything that needs to happen based on https://github.com/dequelabs/axe-core/pull/1836#issuecomment-538650345?

**Comment by AutoSponge** [comment]
@stephenmathieson I assume a code owner of axe-core will need to update the reviewer settings when this file lands.

**Comment by stephenmathieson** [comment]
I think this can land without a "codeowners" update and when @WilcoFiers's back, he can decide what to do moving forward.

**Pull request merged**